subsys/net/buf.c

```diff
@@ -91,23 +79,162 @@ void net_buf_reset(struct net_buf *buf)
 	NET_BUF_ASSERT(buf->flags == 0);
 	NET_BUF_ASSERT(buf->frags == NULL);
 
-	buf->len   = 0;
-	buf->data  = buf->__buf;
+	net_buf_simple_reset(&buf->b);
+}
+
+static u8_t *generic_data_ref(struct net_buf *buf, u8_t *data)
+{
+	u8_t *ref_count;
+
+	ref_count = data - 1;
+	(*ref_count)++;
+
+	return data;
+}
+
+static u8_t *mem_pool_data_alloc(struct net_buf *buf, size_t size,
+				 s32_t timeout)
+{
+	struct net_buf_pool *buf_pool = net_buf_pool_get(buf->pool_id);
+	struct k_mem_pool *pool = buf_pool->alloc->alloc_data;
+	struct k_mem_block block;
+	u8_t *ref_count;
+
+	/* Reserve space for ref-count (u8_t) */
+	size += sizeof(struct k_mem_block_id) + 1;
+	if (k_mem_pool_alloc(pool, &block, size, timeout)) {
+		return NULL;
+	}
+
+	/* save the block descriptor info at the start of the actual block */
+	memcpy(block.data, &block.id, sizeof(block.id));
+
+	ref_count = (u8_t *)block.data + sizeof(block.id);
+	*ref_count = 1;
+
+	/* Return pointer to the byte following the ref count */
+	return ref_count + 1;
+}
+
+static void mem_pool_data_unref(struct net_buf *buf, u8_t *data)
+{
+	struct k_mem_block_id id;
+	u8_t *ref_count;
+
+	ref_count = data - 1;
+	if (--(*ref_count)) {
+		return;
+	}
+
+	/* Need to copy to local variable due to alignment */
+	memcpy(&id, ref_count - sizeof(id), sizeof(id));
+	k_mem_pool_free_id(&id);
+}
+
+const struct net_buf_data_cb net_buf_var_cb = {
+	.alloc = mem_pool_data_alloc,
+	.ref   = generic_data_ref,
+	.unref = mem_pool_data_unref,
+};
+
+static u8_t *fixed_data_alloc(struct net_buf *buf, size_t size, s32_t timeout)
+{
+	struct net_buf_pool *pool = net_buf_pool_get(buf->pool_id);
+	const struct net_buf_pool_fixed *fixed = pool->alloc->alloc_data;
+
+	if (size > fixed->data_size) {
```

Comment:
I don't think that's useful behavior. Let's consider an algorithm to prepare a pkt for sending:

1. Take interface MTU as a basic constraint on the pkt size.
2. Apply additional constraints: connection and protocol specific (e.g. path MTU, TCP's MSS, etc.).
3. Thus, a max pkt size is found.
4. Add protocol headers, decrease the size.
5. Whatever is lesser, input user data size, or remaining pkt size, allocate a data fragment buffer of that size. (Note: given different allocator implementation, a client knows *nothing* about max size a data fragment allocator can handle).
6. Fill in receive data fragment, subtract fragment size from remaining pkt size, if not zero, repeat from 5.


Thus, data buffer allocator would need:

1. Be ready to accept request of arbitrary allocation size.
2. Try to satisfy some of that size (even if smaller). 
3. Allow client to query the size of the allocated buffer, (for step 6 above).

Reply:
This analysis makes very much sense to me. There's already buf->size which indicates the maximum payload size, so what's left to do is to document net_buf_alloc() that it may return a smaller than requested payload if the allocator wasn't able to fulfill the exact request. FWIW, on the Bluetooth side it would in many cases constitute a bug to try to do such an allocation since the pool definition and allocation code are quite tightly coupled. This is the reason I initially implemented this by throwing an error.

Reply:
I've now updated the PR so that the alloc callback is able to allocate less than requested, and the FIXED pool allocator will not fail anymore, but allocate at most the fixed block size.

Reply:
> I've now updated the PR so that the alloc callback is able to allocate less than requested

Thanks, much appreciated!